### PR TITLE
feat: complifi strategy

### PR DESCRIPTION
**Pull request by fp-crypto**
The `ComplifiUSDC` strategy uses USDC to mint equal numbers of fully collateralized 5x leveraged tokens against an underlying (e.g. 2 usdc = 1 5x eth up and 1 5x eth down token). These tokens can be refunded at any time for the same amount of collateral, assuming they are provided in equal numbers. Holding equal numbers of up and down tokens carries no directional risk. Complifi is offering farm rewards in their protocol token, COMFI. The strategy: 

 - Mints 5x leveraged up and down tokens with USDC
 - Stakes them for COMFI rewards
 - Sells the rewards for more USDC
 - Redeems the tokens on demand, by:
     - Withdrawing from the rewards contract
     - Using the vault refund method to redeem equal amounts of up and down tokens for USDC
     
## Notes

We are targeting `yearn-vaults@0.3.0` as USDC prod is on that version.

### CODE REVIEW

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -16,32 +16,106 @@ import {
...
 
     function estimatedTotalAssets() public view override returns (uint256) {
-        // TODO: Build a more accurate estimate using the value of all positions in terms of `want`
-        return want.balanceOf(address(this));
+        uint256 looseTokens =
+            complementToken().balanceOf(address(this)).add(
+                primaryToken().balanceOf(address(this))
+            );
+        (uint256 depositedPrimary, ) =
+            liquidityMining.userPoolInfo(primaryTokenPid(), address(this));
```
Comment (poolpitako) [inline contracts/Strategy.sol]: you can use `liquidityMining.userPoolInfo(primaryTokenPid(), address(this)).amount;`

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -16,32 +16,106 @@ import {
...
+
+    address[] public path;
+
     constructor(address _vault) public BaseStrategy(_vault) {
-        // You can set these parameters on deployment to whatever you want
-        // maxReportDelay = 6300;
-        // profitFactor = 100;
-        // debtThreshold = 0;
+        want.safeApprove(address(tokenVault), type(uint256).max);
+        primaryToken().safeApprove(address(tokenVault), type(uint256).max);
```
Comment (poolpitako) [inline contracts/Strategy.sol]: when you use a method more than once, assign to a local var.

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
...
-        // TODO: Do stuff here to free up any returns back into `want`
-        // NOTE: Return `_profit` which is value generated by all positions, priced in `want`
-        // NOTE: Should try to free up at least `_debtOutstanding` of underlying position
+        liquidityMining.claim();
+        _sell();
+
+        uint256 assets = estimatedTotalAssets();
+        uint256 wantBal = want.balanceOf(address(this));
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
```
Comment (poolpitako) [inline contracts/Strategy.sol:227]: we need to be careful with this when downgrading to 0.3.0

File: contracts/Strategy.sol [hunk 4]
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
...
+
+            _liquidatedAmount = want.balanceOf(address(this));
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
-    // NOTE: Can override `tendTrigger` and `harvestTrigger` if necessary
+    function migrateTokenVault(address _newTokenVault) external onlyGovernance {
+        require(
```
Comment (poolpitako) [inline contracts/Strategy.sol:350]: I think you can kill this. we wouldn't migrate something to a wrong state.

File: contracts/Strategy.sol [hunk 5]
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
...
+            IComplifiVault(_newTokenVault).state() == IComplifiVault.State.Live
+        );
+
+        liquidatePosition(uint256(-1));
```
Comment (poolpitako) [inline contracts/Strategy.sol]: `type(uint256).max`

File: contracts/Strategy.sol [hunk 6]
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
...
-    // Example:
-    //
-    //    function protectedTokens() internal override view returns (address[] memory) {
-    //      address[] memory protected = new address[](3);
-    //      protected[0] = tokenA;
-    //      protected[1] = tokenB;
-    //      protected[2] = tokenC;
-    //      return protected;
-    //    }
+        liquidatePosition(uint256(-1)); //withdraw all. does not matter if we ask for too much
```
Comment (poolpitako) [inline contracts/Strategy.sol]: `type(uint256).max`

File: contracts/Strategy.sol [hunk 7]
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
...
-    // NOTE: Can override `tendTrigger` and `harvestTrigger` if necessary
+    function migrateTokenVault(address _newTokenVault) external onlyGovernance {
+        require(
+            IComplifiVault(_newTokenVault).state() == IComplifiVault.State.Live
+        );
+
+        liquidatePosition(uint256(-1));
+
+        // Revoke approvals for the old token vault
+        want.safeApprove(address(tokenVault), type(uint256).min);
```
Comment (poolpitako) [inline contracts/Strategy.sol]: `0` instead of `type(uint256).min`?

File: contracts/Strategy.sol (hunk 2 above)
Comment (fp-crypto) [reply contracts/Strategy.sol]: do you mean `address(tokenVault)`?

File: contracts/Strategy.sol (hunk 3 above)
Comment (fp-crypto) [reply contracts/Strategy.sol:227]: what changes?

File: contracts/Strategy.sol (hunk 2 above)
Comment (poolpitako) [reply contracts/Strategy.sol]:
```
address primaryToken = primaryToken();

primaryToken.firstUsage();
primaryToken.secondUsage();
```

File: contracts/Strategy.sol (hunk 7 above)
Comment (fp-crypto) [reply contracts/Strategy.sol]: Moved approvals to a helper because DRY

**Pull request merged**